File: management-domain/Executive-domain-X/Executive-domain-X/Backend/Nodes/executive-command-service/src/models/KPI.js
/**
 * KPI Domain Model
 */

const { ObjectId } = require('mongodb');

/**
 * KPI (Key Performance Indicator) Model
 * Represents a high-level metric tracked by executives
 */
class KPI {
  constructor(data) {
    this._id = data._id || new ObjectId();
    this.tenantId = data.tenantId;
    this.name = data.name;
    this.category = data.category; // FINANCIAL, OPERATIONAL, CUSTOMER, EMPLOYEE
    this.executiveLevel = data.executiveLevel || 'ALL'; // CEO, CFO, CTO, COO, ALL
    this.value = data.value;
    this.unit = data.unit; // $, %, count, score
    this.period = data.period; // 2024-01, Q1-2024, YTD-2024
    this.target = data.target;
    this.previousValue = data.previousValue;
    this.percentChange = data.percentChange;
    this.status = data.status || 'ON_TRACK'; // ON_TRACK, AT_RISK, BEHIND, AHEAD
    this.trend = data.trend; // UP, DOWN, STABLE
    this.dataSources = data.dataSources || [];
    this.metadata = data.metadata || {};
    this.visible = data.visible !== false;
    this.isCalculated = data.isCalculated !== false;
    this.lastCalculatedAt = data.lastCalculatedAt || new Date();
    this.createdAt = data.createdAt || new Date();
    this.updatedAt = data.updatedAt || new Date();
  }

  /**
   * Convert to MongoDB document
   */
  toDocument() {
    return {
      _id: this._id,
      tenantId: this.tenantId,
      name: this.name,
      category: this.category,
      executiveLevel: this.executiveLevel,
      value: this.value,
      unit: this.unit,
      period: this.period,
      target: this.target,
      previousValue: this.previousValue,
      percentChange: this.percentChange,
      status: this.status,
      trend: this.trend,
      dataSources: this.dataSources,
      metadata: this.metadata,
      visible: this.visible,
      isCalculated: this.isCalculated,
      lastCalculatedAt: this.lastCalculatedAt,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
    };
  }

  /**
   * Create KPI from MongoDB document
   */
  static fromDocument(doc) {
    if (!doc) return null;
    return new KPI({
      _id: doc._id,
      tenantId: doc.tenantId,
      name: doc.name,
      category: doc.category,
      executiveLevel: doc.executiveLevel,
      value: doc.value,
      unit: doc.unit,
      period: doc.period,
      target: doc.target,
      previousValue: doc.previousValue,
      percentChange: doc.percentChange,
      status: doc.status,
      trend: doc.trend,
      dataSources: doc.dataSources || [],
      metadata: doc.metadata || {},
      visible: doc.visible,
      isCalculated: doc.isCalculated,
      lastCalculatedAt: doc.lastCalculatedAt,
      createdAt: doc.createdAt,
      updatedAt: doc.updatedAt,
    });
  }

  /**
   * Calculate percent change from previous value
   */
  calculatePercentChange() {
    if (this.previousValue && this.previousValue !== 0) {
      const change = this.value - this.previousValue;
      this.percentChange = (change / this.previousValue) * 100;
    } else {
      this.percentChange = null;
    }
    return this.percentChange;
  }

  /**
   * Update status based on value vs target
   */
  updateStatus() {
    if (this.target && this.target !== 0) {
      const ratio = this.value / this.target;

      if (ratio >= 1.1) {
        this.status = 'AHEAD';
        this.trend = 'UP';
      } else if (ratio >= 0.9) {
        this.status = 'ON_TRACK';
        this.trend = 'STABLE';
      } else if (ratio >= 0.8) {
        this.status = 'AT_RISK';
        this.trend = 'DOWN';
      } else {
        this.status = 'BEHIND';
        this.trend = 'DOWN';
      }
    }
    return this.status;
  }

  /**
   * Check if KPI is on track
   */
  isOnTrack() {
    return this.status === 'ON_TRACK' || this.status === 'AHEAD';
  }

  /**
   * Check if KPI needs attention
   */
  needsAttention() {
    return this.status === 'AT_RISK' || this.status === 'BEHIND';
  }

  /**
   * Add metadata
   */
  addMetadata(key, value) {
    this.metadata[key] = value;
    return this;
  }

  /**
   * Get collection name
   */
  static getCollectionName() {
    return 'kpi_metrics';
  }

  /**
   * Get index specifications
   */
  static getIndexes() {
    return [
      { key: { tenantId: 1, category: 1, period: -1 } },
      { key: { tenantId: 1, executiveLevel: 1, category: 1, period: -1 } },
      { key: { tenantId: 1, period: -1 } },
      { key: { tenantId: 1, name: 1 } },
    ];
  }
}

module.exports = KPI;

File: management-domain/Executive-domain-X/Executive-domain-X/Backend/Java/executive-dashboard-service/Shared/executive-analytics-service/src/main/java/com/gogidix/management/executive/analytics/domain/model/KPI.java
package com.gogidix.management.executive.analytics.domain.model;

import com.gogidix.management.shared.domain.BaseEntity;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.TypeAlias;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Document(collection = "kpi_metrics")
@TypeAlias("kpi")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = true)
@CompoundIndex(name = "kpi_tenant_category_idx", def = "{'tenantId': 1, 'category': 1, 'period': -1}")
@CompoundIndex(name = "kpi_tenant_executive_idx", def = "{'tenantId': 1, 'executiveLevel': 1, 'category': 1, 'period': -1}")
public class KPI extends BaseEntity {

    @Indexed
    private String name;
    private String category;
    private String executiveLevel;
    private BigDecimal value;
    private String unit;
    private String period;
    private BigDecimal target;
    private BigDecimal previousValue;
    private BigDecimal percentChange;
    private String status;
    private String trend;
    private List<String> dataSources;
    private Map<String, Object> metadata;
    @Builder.Default
    private Boolean visible = true;
    private Instant lastCalculatedAt;
    private Boolean isCalculated = true;

    public KPI(String tenantId, String name, String category, BigDecimal value) {
        super(tenantId);
        this.name = name;
        this.category = category;
        this.value = value;
        this.lastCalculatedAt = Instant.now();
    }

    public KPI(String tenantId, String name, String category, String executiveLevel, BigDecimal value, String period) {
        super(tenantId);
        this.name = name;
        this.category = category;
        this.executiveLevel = executiveLevel;
        this.value = value;
        this.period = period;
        this.lastCalculatedAt = Instant.now();
    }

    public boolean isOnTrack() {
        return "ON_TRACK".equals(this.status) || "AHEAD".equals(this.status);
    }

    public boolean needsAttention() {
        return "AT_RISK".equals(this.status) || "BEHIND".equals(this.status);
    }

    public void calculatePercentChange() {
        if (previousValue != null && previousValue.compareTo(BigDecimal.ZERO) != 0 && value != null) {
            BigDecimal change = value.subtract(previousValue);
            this.percentChange = change.divide(previousValue, 4, RoundingMode.HALF_UP)
                    .multiply(new BigDecimal("100"));
        }
    }

    public void updateStatus() {
        if (target != null && target.compareTo(BigDecimal.ZERO) != 0 && value != null) {
            BigDecimal ratio = value.divide(target, 2, RoundingMode.HALF_UP);
            if (ratio.compareTo(new BigDecimal("1.1")) >= 0) {
                this.status = "AHEAD";
                this.trend = "UP";
            } else if (ratio.compareTo(new BigDecimal("0.9")) >= 0) {
                this.status = "ON_TRACK";
                this.trend = "STABLE";
            } else if (ratio.compareTo(new BigDecimal("0.8")) >= 0) {
                this.status = "AT_RISK";
                this.trend = "DOWN";
            } else {
                this.status = "BEHIND";
                this.trend = "DOWN";
            }
        }
    }

    public void markAsRecalculated() {
        this.lastCalculatedAt = Instant.now();
        this.touch();
    }

    public void addDimension(String key, String value) {
        if (this.metadata == null) {
            this.metadata = new HashMap<>();
        }
        this.metadata.put("dimension_" + key, value);
    }

    public String getDimension(String key) {
        if (this.metadata == null) {
            return null;
        }
        Object value = this.metadata.get("dimension_" + key);
        return value != null ? value.toString() : null;
    }

    public Map<String, Object> getDimensions() {
        if (this.metadata == null) {
            return new HashMap<>();
        }
        Map<String, Object> dimensions = new HashMap<>();
        for (Map.Entry<String, Object> entry : this.metadata.entrySet()) {
            if (entry.getKey().startsWith("dimension_")) {
                dimensions.put(entry.getKey().substring("dimension_".length()), entry.getValue());
            }
        }
        return dimensions;
    }

    public void addMetadata(String key, Object value) {
        if (this.metadata == null) {
            this.metadata = new HashMap<>();
        }
        this.metadata.put(key, value);
    }

    public Object getMetadata(String key) {
        if (this.metadata == null) {
            return null;
        }
        return this.metadata.get(key);
    }
}
